### PR TITLE
Fix measure performance frequency + Add loss log

### LINKED ISSUE #53: [BUG] - Measuring performance at the end of each epoch, disables training logging
The title says pretty much all.

If I want to measure the performance of the model only at the end of every epoch, the (for example) ClassifierTrainer offers me the possibility of setting a value for the logging frequency to something `<= 0`.

In this way, I measure the performance at the end of every epoch on the validation set (it works).

However, on tensorboard I only see the plot of the validation curves, the training curves aren't displayed anymore

### LINKED ISSUE #57: [BUG] - Restorer does not work with GANs conditioned
**Describe the bug**
Run `examples/gans/facades.py`, stop the training after some time and then try to restart the training. You get an error due to the fact that the discriminator expects a list of two tensors.

**Expected behavior**
The training should restart with models restored correctly.

**Code to reproduce the issue**
Simply run `examples/gans/facades.py`.

**Other info / logs**
Inside `build_and_restore` we should check the size of discriminator inputs (and also the size of generator inputs).

### LINKED ISSUE #58: [BUG] - Restorer does not work with distributed training
**Describe the bug**
Run `examples/gan/pix2pix_facades_multi_gpu.py` in a multi-gpu scenario. If you try to restore the training once finished you get an error due to wrong input shapes.
This is because in a multi-gpu scenario the batch size gets updated based on the number of devices.
Simply move the call to `build_or_restore` after the batch size update.

**Expected behavior**
The restorer should restore the models.

**Code to reproduce the issue**
`examples/gan/pix2pix_facades_multi_gpu.py`

### LINKED ISSUE #60: [BUG/API] - Passing name to Custom ClassifierMetric(s)
**System information**
- AshPy version:

**Describe the bug**
No easy way to specify a name for subclassed custom `ClassifierMetric`

**Expected behavior**
While subclassing `ClassifierMetric` it should be possible to specify a name for the metric.

### LINKED ISSUE #61: [FEATURE] - Add a flag for correct scaling of logged TensorBoard images
**System information**
- AshPy version (you are using): 0.4.0
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state**
TensorBoard images should be logged in the [0, 1] range when float.
Currently, we do not provide any type of automatic handling of this scaling, so if you are training on [-1, 1] and using our callbacks, images will be logged without scaling.
 
**Will this change the current API? How?**
Small change to `ashpy.utils.log()` and all the callbacks using it, they will now accept a new boolean argument `auto_tb_scaling` controlling this behavior.

**Who will benefit with this feature?**
Anyone using the default logging facilities.

**Any Other info**

### PR DESCRIPTION
**Pull request by EmanueleGhelfi**
## Description

Fixes #58 
Fixes #57 
Fixes #53 
Fixes #60 

Fixes examples (they were broken)

## Type of change

Please delete options that are not relevant.

- [x] Bugfix (i.e., a non-breaking change which fixes an issue)
- [x] New feature (i.e., a non-breaking change which adds functionality)
- [ ] Breaking change (i.e., a fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by EmanueleGhelfi** [comment]
Work in Progress

File: src/ashpy/callbacks/classifier.py [hunk 1]
```diff
@@ -32,24 +32,27 @@ def __init__(
...
         Args:
+            name: name of the callback
             event: event to consider
             event_freq: frequency of logging
-
+            input_is_zero_centered: if True, the callback assumes the input is in [-1, 1] if it
+                is an image with type tf.float. If False, the callback assumes the input is [0,
+                1] if type float, and [0, 255] if type is uint. If the input type is float and
+                the image is in [0, 1] use False. If the input type is uint this parameter is
+                ignored.
```
Comment (mr-ubik) [inline src/ashpy/callbacks/classifier.py:48]: Please add Type Annotations to both signature and docs

File: src/ashpy/callbacks/classifier.py [hunk 2]
```diff
@@ -60,5 +63,15 @@ def _log_fn(context: ClassifierContext) -> None:
         input_tensor = context.current_batch[0]
         out_label = context.current_batch[1]
 
+        rank = tf.rank(input_tensor)
+
+        # if it is an image scale it if needed
```
Comment (mr-ubik) [inline src/ashpy/callbacks/classifier.py]: `# If it is an image scale it is needed`

File: src/ashpy/losses/classifier.py [hunk 3]
```diff
@@ -24,7 +24,7 @@
 class ClassifierLoss(Executor):
     r"""Classifier Loss Executor using the classifier model, instantiated with a fn."""
 
-    def __init__(self, fn: tf.keras.losses.Loss) -> None:
+    def __init__(self, fn: tf.keras.losses.Loss, name="ClassifierLoss") -> None:
```
Comment (mr-ubik) [inline src/ashpy/losses/classifier.py]: Please add Type Annotations to both signature and docs

File: src/ashpy/losses/executor.py [hunk 4]
```diff
@@ -29,7 +29,7 @@
 class Executor:
     """Carry a function and the way of executing it. Given a context."""
 
-    def __init__(self, fn: tf.keras.losses.Loss = None) -> None:
+    def __init__(self, fn: tf.keras.losses.Loss = None, name="loss") -> None:
```
Comment (mr-ubik) [inline src/ashpy/losses/executor.py]: Please add Type Annotations to both signature and docs

File: src/ashpy/callbacks/classifier.py [hunk 5]
```diff
@@ -60,5 +63,15 @@ def _log_fn(context: ClassifierContext) -> None:
...
+        if (
+            tf.equal(rank, 4)
```
Comment (mr-ubik) [inline src/ashpy/callbacks/classifier.py:70]: Are you sure that images will always have rank == 4? I may be mistaken, but I remember that some libraries read grayscale as [x, y] and not [x, y, c]

File: src/ashpy/losses/executor.py [hunk 6]
```diff
@@ -153,17 +155,29 @@ def __call__(self, context, **kwargs) -> tf.Tensor:
...
 
         """
-        return self._weight(context.global_step) * self.call(context, **kwargs)
+        self._loss_value = self._weight(context.global_step) * self.call(
+            context, **kwargs
+        )
+        return self._loss_value
 
-    def __add__(self, other) -> SumExecutor:
+    def log(self, step):
```
Comment (mr-ubik) [inline src/ashpy/losses/executor.py]: Please add Type Annotations to both signature and docs

File: src/ashpy/losses/executor.py [hunk 7]
```diff
@@ -185,7 +199,7 @@ def __mul__(self, other: Union[Callable[..., float], float, int, tf.Tensor]):
             self._weight = lambda step: weight(step) * __other(step)
         return self
 
-    def __rmul__(self, other):
+    def __rmul__(self, other: Union[SumExecutor, Executor]):
         """See `__mul__` method."""
```
Comment (mr-ubik) [inline src/ashpy/losses/executor.py]: ```__mul__``` will actually be converted as _\_\_mul\_\__ . Here use one of the following:

```rst
``__mul__``
:code:`__mul__`
:meth:`ashpy.losses.Executor.__mul__() <ashpy.losses.executor.Executor.__mul__()>`
```
The last one is probably the best one.

File: src/ashpy/losses/executor.py [hunk 8]
```diff
@@ -198,7 +212,7 @@ class SumExecutor(Executor):
     then summed together.
     """
 
-    def __init__(self, executors) -> None:
+    def __init__(self, executors: List[Executor], name: str = "LossSum") -> None:
```
Comment (mr-ubik) [inline src/ashpy/losses/executor.py:221]: Add `name` to docs

File: src/ashpy/losses/executor.py [hunk 9]
```diff
@@ -235,8 +249,21 @@ def call(self, *args, **kwargs) -> tf.Tensor:
...
 
         """
-        result = tf.add_n([executor(*args, **kwargs) for executor in self._executors])
-        return result
+        self._loss_value = tf.add_n(
+            [executor(*args, **kwargs) for executor in self._executors]
+        )
+        return self._loss_value
+
+    def log(self, step):
```
Comment (mr-ubik) [inline src/ashpy/losses/executor.py]: Please add Type Annotations to both signature and docs

File: src/ashpy/losses/gan.py [hunk 10]
```diff
@@ -114,7 +114,9 @@ def get_discriminator_inputs(
 class GeneratorAdversarialLoss(GANExecutor):
     r"""Base class for the adversarial loss of the generator."""
 
-    def __init__(self, loss_fn: tf.keras.losses.Loss = None) -> None:
+    def __init__(
+        self, loss_fn: tf.keras.losses.Loss = None, name="GeneratorAdversarialLoss"
```
Comment (mr-ubik) [inline src/ashpy/losses/gan.py]: Please add Type Annotations to both signature and docs

Comment (mr-ubik) [reply src/ashpy/losses/gan.py]: Also very important: must this `name` argument be unique? Does it behave like the `name` argument for `Metric` and `Callback`?

File: src/ashpy/losses/gan.py [hunk 11]
```diff
@@ -515,7 +518,9 @@ def call(
 class DiscriminatorAdversarialLoss(GANExecutor):
     r"""Base class for the adversarial loss of the discriminator."""
 
-    def __init__(self, loss_fn: tf.keras.losses.Loss = None) -> None:
+    def __init__(
```
Comment (mr-ubik) [inline src/ashpy/losses/gan.py:525]: Please add Type Annotations to both signature and docs

File: src/ashpy/metrics/classifier.py [hunk 12]
```diff
@@ -90,6 +90,7 @@ class ClassifierMetric(Metric):
     def __init__(
         self,
         metric: tf.keras.metrics.Metric,
+        name: str = None,
```
Comment (mr-ubik) [inline src/ashpy/metrics/classifier.py]: I know it is implicit but maybe we should make the `Optional` explicit

**Review by mr-ubik — CHANGES_REQUESTED** [review]
Mostly my OCD about type hints and docs + two clarifications

File: src/ashpy/callbacks/classifier.py (hunk 5 above)
Comment (EmanueleGhelfi) [reply src/ashpy/callbacks/classifier.py:70]: We need to check this. We have the same code inside log/utils.py (function log) so we maybe need to fix both of them.

File: src/ashpy/losses/gan.py (hunk 10 above)
Comment (EmanueleGhelfi) [reply src/ashpy/losses/gan.py]: Yes, the loss name should be unique. We actually log the loss as ashpy/loss_name.

Comment (EmanueleGhelfi) [reply src/ashpy/losses/gan.py]: Where can we specify the fact that the name should be unique?

File: src/ashpy/callbacks/classifier.py (hunk 5 above)
Comment (galeone) [reply src/ashpy/callbacks/classifier.py:70]: For TensorFlow images must have always rank 3. A batch of images rank 4.

@mr-ubik opencv reads grayscale images without the depth channel (WxH) but for using them with TF we **must** add the depth channel, otherwise any tf.image* transformation won't work.

In short, this is correct

File: src/ashpy/losses/gan.py (hunk 10 above)
Comment (galeone) [reply src/ashpy/losses/gan.py]: In the documentation + we must keep track of every loss passed to a trainer, store the names and check for duplicates. If there are duplicates, raise an exception

Comment (EmanueleGhelfi) [reply src/ashpy/losses/gan.py]: Add check in trainers, need to add documentation.
